Add `profiling` feature which uses `info_span!` for code profiling

heavyrain266 opened the pull request:
Follows #20

heavyrain266 committed:
Add profiling feature which uses info_span! for code profiling

Follows #20

nicopap commented on src/loader.rs:
```diff
@@ -311,12 +329,24 @@ impl<'b, 'w> Loader<'b, 'w> {
 
                 let label = format!("{label}{i}");
-                trace!("Successfully loaded geometry mesh: {label}");
+
+                {
+                    let loaded = trace_span!("Loaded geometry mesh").entered();
+                    trace!("Successfully loaded geometry mesh: {label}");
+
+                    drop(loaded)
+                }
```
```suggestion
let _load_geometry_span = trace_span!("Load Geometry mesh", label = &label).entered();
```

nicopap reviewed (changes requested):
Tracing spans should encompass all the code it measures.

A `span` denotes a length of code that you want to measure the runtime of.

![tracy screenshot](https://user-images.githubusercontent.com/302146/163988636-25c017ab-64bc-4da7-a897-a80098b667ef.png)

In the picture, each colored bar is a tracing span.

Here, you are just measuring how much time it takes to log a value.

To trace how long it takes to do something, you would need to create the `span` object by calling `let load_mesh_span = span!(...)` **before** the computation; Then, call `drop(load_mesh_span)` **after** it is done.

nicopap reviewed (changes requested):
This is much better, but two points still need attention:
1. I think it's better to replicate how bevy does tracing (simply because I've no experience tracing outside of bevy):
  a. Replace `trace_span!` by `info_span!`
  b. Create a `trace` cargo feature and use `#[cfg(feature="trace")]` wherever `info_span!` is used [see the bevy code](https://github.com/bevyengine/bevy/blob/c6a41cdd10c77b1e563efde2b721eb364e4f3143/crates/bevy_render/src/lib.rs#L210-L239)
  c. You would need to add the two lines `[features]\ntrace = []` into the `Cargo.toml`.
2. Avoid using `span` for logging information. The input argument of the `info_span!` macro should **describe what the code is doing** during the span. I think it's OK to log the `label`, because it says on what the code is operating, it could be useful for users trying to optimize their own assets. But, if you want to log information like the number of vertices, I suggest you use `trace!` or `debug!` directly instead of a span.
3. Maybe add more spans around expensive operations (like triangulation)

### Here is a sample to get you started

Instead of the following:
```rust
        let _load_geometry_span = trace_span!("Load Geometry mesh", label = &label).entered();

        let polygon_vertices = mesh_obj
            .polygon_vertices()
            .context("Failed to get polygon vertices")?;
        let triangle_pvi_indices = polygon_vertices
            .triangulate_each(triangulate::triangulate)
            .context("Triangulation failed")?;
       // ...
       let _expand_position_len_span =
            trace_span!("Expand position lenght", len = &positions.len()).entered();
       // rest of the load_bevy_mesh method
```
you would write that:
```rust
        #[cfg(feature = "trace")]
        let _load_geometry_span = info_span!("load_geometry_mesh", label = &label).entered();

        #[cfg(feature = "trace")]
        let triangulate_mesh_span = info_span!("triangulate", label = &label).entered();

        let polygon_vertices = mesh_obj
            .polygon_vertices()
            .context("Failed to get polygon vertices")?;
        let triangle_pvi_indices = polygon_vertices
            .triangulate_each(triangulate::triangulate)
            .context("Triangulation failed")?;
        #[cfg(feature = "trace")]
        drop(triangulate_mesh_span);
       // ...
        trace!("Expand position lenght {}", positions.len());
       // rest of the load_bevy_mesh method
```

Here is a screenshot of tracy with the modifications I suggested:

![_Ascreencap](https://user-images.githubusercontent.com/26321040/181220685-f5df93eb-4594-4445-89ad-435db4fb9468.png)

nicopap commented on src/loader.rs:
```diff
@@ -183,7 +184,9 @@ impl<'b, 'w> Loader<'b, 'w> {
-        trace!("Expanded positions len: {:?}", positions.len());
+
+        let _expand_position_len_span =
+            trace_span!("Expand position lenght", len = &positions.len()).entered();
```
This would better be replaced by `debug!(..)`

nicopap commented on src/loader.rs:
```diff
@@ -312,11 +317,14 @@ impl<'b, 'w> Loader<'b, 'w> {
                 let label = format!("{label}{i}");
-                trace!("Successfully loaded geometry mesh: {label}");
+
+                let _loaded_geometry_mesh_span =
+                    trace_span!("Successfully loaded geometry mesh", label = &label).entered();
```
```suggestion
```

This doesn't correspond to a "span"

nicopap commented on src/loader.rs:
```diff
@@ -288,7 +291,9 @@ impl<'b, 'w> Loader<'b, 'w> {
         };
-        trace!("material count for this mesh: {}", all_indices.len());
+
+        let _material_count_per_mesh =
+            trace_span!("Material count for this mesh", count = all_indices.len()).entered();
```
This would better be replaced by `debug!(..)`

nicopap commented on src/loader.rs:
```diff
@@ -371,7 +379,9 @@ impl<'b, 'w> Loader<'b, 'w> {
             .set_labeled_asset(&label, LoadedAsset::new(mesh.clone()));
-        trace!("Successfully loaded FBX mesh: {label}");
+
+        let _loaded_fbx_mesh =
+            trace_span!("Successfully loaded FBX mesh", label = &label).entered();
```
```suggestion
```

This doesn't correspond to a "span"

nicopap committed:
Add a maya PBR material loader (#23)

* Add a maya PBR material loader

The fields to use were reverse-engineered from a file using maya's PBR
material FBX extension.

* Add precision on maya loader assumptions

Replace info! with debug!

nicopap commented on src/loader.rs:
```diff
@@ -485,7 +498,7 @@ impl<'b, 'w> Loader<'b, 'w> {
             // for each image or texture to load.
             let handle = if let Some(handle) = self.scene.textures.get(&handle_label) {
-                trace!("Already encountered texture: {label}, skipping");
+                warn!("Already encountered texture: {label}, skipping");
```
```suggestion
                debug!("Already encountered texture: {label}, skipping");
```

This shouldn't be a warn, it's expected. We cache the textures to avoid spending time reloading them

nicopap commented on src/loader.rs:
```diff
@@ -552,12 +565,12 @@ impl<'b, 'w> Loader<'b, 'w> {
         };
         if let Some(handle) = self.scene.materials.get(&label) {
-            trace!("Already encountered material: {label}, skipping");
+            warn!("Already encountered material: {label}, skipping");
```
```suggestion
            debug!("Already encountered material: {label}, skipping");
```

same

nicopap commented on src/loader.rs:
```diff
@@ -552,12 +565,12 @@ impl<'b, 'w> Loader<'b, 'w> {
         }
 
-        trace!("Loading material: {label}");
+        info!("Loading FBX material: {label}");
```
```suggestion
        debug!("Loading FBX material: {label}");
```
I'd rather not enable it on `info`, this seems too verbose.

nicopap commented on src/loader.rs:
```diff
@@ -572,7 +585,7 @@ impl<'b, 'w> Loader<'b, 'w> {
             .set_labeled_asset(&label, LoadedAsset::new(material));
 
-        trace!("Successfully loaded material: {label}");
+        info!("Successfully loaded material: {label}");
```
```suggestion
        debug!("Successfully loaded material: {label}");
```

same here

nicopap commented on src/loader.rs:
```diff
@@ -391,13 +405,12 @@ impl<'b, 'w> Loader<'b, 'w> {
         video_clip_obj: object::video::ClipHandle<'_>,
     ) -> anyhow::Result<Image> {
-        trace!("Loading texture image: {:?}", video_clip_obj.name());
+        info!("Loading texture image: {:?}", video_clip_obj.name());
```
```suggestion
        debug!("Loading texture image: {:?}", video_clip_obj.name());
```

This is too verbose, info log level is visible by default. This only serves to check if the function ran, and the message is printed a lot. At most, I'd put it to `debug` level.

nicopap commented on src/loader.rs:
```diff
@@ -425,7 +438,7 @@ impl<'b, 'w> Loader<'b, 'w> {
         let image = image.context("Failed to read image buffer data")?;
 
-        trace!(
+        info!(
```
```suggestion
        debug!(
```

same here

nicopap commented on src/loader.rs:
```diff
@@ -380,8 +395,7 @@ impl<'b, 'w> Loader<'b, 'w> {
 
-        let _loaded_fbx_mesh =
-            trace_span!("Successfully loaded FBX mesh", label = &label).entered();
+        info!("Successfully loaded FBX mesh: {label}");
```
```suggestion
        debug!("Successfully loaded FBX mesh: {label}");
```

nicopap commented on src/loader.rs:
```diff
@@ -316,15 +333,14 @@ impl<'b, 'w> Loader<'b, 'w> {
 
-                let _loaded_geometry_mesh_span =
-                    trace_span!("Successfully loaded geometry mesh", label = &label).entered();
+                info!("Successfully loaded geometry mesh: {label}");
```
```suggestion
```
can be removed now that we have spans.

nicopap commented on src/loader.rs:
```diff
@@ -160,8 +171,13 @@ impl<'b, 'w> Loader<'b, 'w> {
             Some(name) if !name.is_empty() => format!("FbxMesh@{name}/Primitive"),
             _ => format!("FbxMesh{}/Primitive", mesh_obj.object_id().raw()),
         };
+        info!("Loading geometry mesh: {label}");
```
```suggestion
```

can be removed now that we have spans.

nicopap commented on src/loader.rs:
```diff
@@ -347,8 +363,7 @@ impl<'b, 'w> Loader<'b, 'w> {
         };
-
-        trace!("Loading mesh: {label}");
+        info!("Loading FBX mesh: {label}");
```
```suggestion
        debug!("Loading FBX mesh: {label}");
```
This is too verbose, info log level is visible by default. This only serves to check if the function ran, and the message is printed a lot. At most, I'd put it to debug level.

nicopap reviewed (changes requested):
Once the log levels are adjusted, this is good to go.

heavyrain266 committed:
Merge branch 'main' into use-spans

nicopap reviewed (approved):
LGTM
